blas/gonum/level1float64.go

```diff
@@ -237,43 +237,78 @@ func (Implementation) Daxpy(n int, alpha float64, x []float64, incX int, y []flo
 	f64.AxpyInc(alpha, x, y, uintptr(n), uintptr(incX), uintptr(incY), uintptr(ix), uintptr(iy))
 }
 
-// Drotg computes the plane rotation
-//   _    _      _ _       _ _
-//  |  c s |    | a |     | r |
-//  | -s c |  * | b |   = | 0 |
-//   ‾    ‾      ‾ ‾       ‾ ‾
-// where
-//  r = ±√(a^2 + b^2)
-//  c = a/r, the cosine of the plane rotation
-//  s = b/r, the sine of the plane rotation
+// Drotg computes a plane rotation
+//  ⎡  c s ⎤ ⎡ a ⎤ = ⎡ r ⎤
+//  ⎣ -s c ⎦ ⎣ b ⎦   ⎣ 0 ⎦
+// satisfying c^2 + s^2 = 1.
 //
-// NOTE: There is a discrepancy between the reference implementation and the BLAS
-// technical manual regarding the sign for r when a or b are zero.
-// Drotg agrees with the definition in the manual and other
-// common BLAS implementations.
+// The computation uses the formulas
+//  sigma = sgn(a)    if |a| >  |b|
+//        = sgn(b)    if |b| >= |a|
+//  r = sigma*sqrt( a^2 + b^2 )
+//  c = 1; s = 0      if r = 0
+//  c = a/r; s = b/r  if r != 0
+//  c >= 0            if |a| > |b|
+//
+// The subroutine also computes
+//  z = s    if |a| > |b|,
+//    = 1/c  if |b| >= |a| and c != 0
+//    = 1    if c = 0
+// This allows c and s to be reconstructed from z as follows:
+//  If z = 1, set c = 0, s = 1.
+//  If |z| < 1, set c = sqrt(1 - z^2) and s = z.
+//  If |z| > 1, set c = 1/z and s = sqrt( 1 - c^2).
+//
+// NOTE: There is a discrepancy between the reference implementation and the
+// BLAS technical manual regarding the sign for r when a or b are zero. Drotg
+// agrees with the definition in the manual and other common BLAS
+// implementations.
 func (Implementation) Drotg(a, b float64) (c, s, r, z float64) {
-	if b == 0 && a == 0 {
-		return 1, 0, a, 0
-	}
-	absA := math.Abs(a)
-	absB := math.Abs(b)
-	aGTb := absA > absB
-	r = math.Hypot(a, b)
-	if aGTb {
-		r = math.Copysign(r, a)
-	} else {
-		r = math.Copysign(r, b)
-	}
-	c = a / r
-	s = b / r
-	if aGTb {
-		z = s
-	} else if c != 0 { // r == 0 case handled above
-		z = 1 / c
-	} else {
+	// Implementation based on Supplemental Material to:
+	// Edward Anderson. 2017. Algorithm 978: Safe Scaling in the Level 1 BLAS.
+	// ACM Trans. Math. Softw. 44, 1, Article 12 (July 2017), 28 pages.
+	// DOI: https://doi.org/10.1145/3061665
+
+	const safmin = 0x1p-1022
+	const safmax = 1 / safmin
+
+	anorm := math.Abs(a)
+	bnorm := math.Abs(b)
+	switch {
+	case bnorm == 0:
```

Comment:
I put this case first to preserve what Drotg returns for `a=b=0`. I'm at the same time asking the reference in https://github.com/Reference-LAPACK/lapack/issues/557 whether they would consider restoring the previous return values, otherwise we have to adapt (the code here and the test expectation).

Reply:
I think it's worth adding a comment to this effect if that issue is not resolved before this is merged.

Reply:
Yes, good idea. Done.

Reply:
The Reference has made the same change in https://github.com/Reference-LAPACK/lapack/pull/560, so I've removed the comment.